onnxruntime/core/providers/cpu/ml/svmclassifier.h

```diff
@@ -30,31 +30,30 @@ class SVMCommon {
   KERNEL get_kernel_type() const { return kernel_type_; }
 
   float kernel_dot(const T* A, int64_t a, const std::vector<float>& B, int64_t b, int64_t len, KERNEL k) const {
-    float sum = 0.f;
+    double sum = 0;
+    const T* pA = A + a;
+    const float* pB = &B[b];
     if (k == KERNEL::POLY) {
-      for (int64_t i = 0; i < len; i++) {
-        sum += B[b + i] * static_cast<float>(A[a + i]);
-      }
+      for (int64_t i = len; i > 0; --i, ++pA, ++pB)
+        sum += *pA * *pB;
       sum = gamma_ * sum + coef0_;
       sum = std::pow(sum, degree_);
     } else if (k == KERNEL::SIGMOID) {
-      for (int64_t i = 0; i < len; i++) {
-        sum += B[b + i] * static_cast<float>(A[a + i]);
-      }
+      for (int64_t i = len; i > 0; --i, ++pA, ++pB)
+        sum += *pA * *pB;
```

Comment:
>*pA [](start = 15, length = 3)

Cast to float?

Reply:
The compiler should be able to choose the right cast. I added a unit test to check that float and double inputs are working.